## Experimentos
Copiamos los códigos de las funciones usadas en QR, e importamos las matrices propuestas para usar como ejemplos

In [3]:
import numpy as np
def QR_factor(A): #Gram-Schmidt modificado
    """
    Entrada: A matriz de numpy invertible n x n
    Proceso de ortogonalización de Gram-Smith modificado en los vectores columna de A
    Salida: Q y R con Q ortogonal y R triangular superior con diagonal positiva tales que A = Q R
    """
    n = A.shape[0]
    v = np.zeros((n,n)) 
    Q = np.zeros((n,n))
    R = np.zeros((n,n))
   
    for i in range(n):
        v[:,i] = A[:,i] #Inicializamos A
    #print(v)
    for i in range(n):
        R[i,i] = np.linalg.norm( v[:,i]) #Como esto es el coeficiente de q_i en la i-ésima fila, es la i,i-ésima componente de R
        Q[:,i] = (1/R[i,i]) * v[:,i] #Normalizamos v_i
        for k in range(i+1,n):
            R[i,k] = Q[:,i] @ v[:,k] #Esto será el coeficiente de q_i en la k-ésima fila, por tanto es la i,k-ésima componente de R
            v[:,k] = v[:,k] - R[i,k] * Q[:,i] #Actualizamos v_k restándole la proyección <q_i,v_k>q_i
    return Q,R

In [41]:
def bottom_half(A):
    """
    Almacena las componentes debajo de la diagonal superior de A
    """
    n = A.shape[0]
    L=[]
    for i in range(n):
        T = list(A[i,:])
        for k in range(i):   
            L.append(T[k])
    return L

In [99]:
def QR_algor(A,max_iteraciones=10000,tol=10e-8):
    """
    Entrada: A matriz de numpy que debería ser simple e invertible, con autovalores distintos en módulo
             max_iteraciones: tope de recursion. si se alcanza se retorna error
             tol: tolerancia, que tan cerca de ser triangular superior es la matriz A^(n) en el paso n
    Salida: B matriz (casi) triangular superior similar a A, con sus autovalores ordenados por su módulo en la diagonal, lista con los autovalores, iteraciones necesarias
    """
    iteracion = 0
    n=A.shape[0]
    while iteracion < max_iteraciones:
        iteracion += 1
        Q,R = QR_factor(A) #Calculamos la ddescomposición A_k = Q'_k R'_k
        A = R @ Q  #Y hacemos A_(k+1) = R'_k Q'_k. Sabemos que sto es equivalente a Q_k^T A Q_k obtenidas por iteración simultánea, y converge a una matriz triangular superior
        L = [abs(l) for l in bottom_half(A)] #Para chequear si estamos suficientemente cerca, tomamos los valores absolutos de la mitad triangular inferior de A...
        if max(L) < tol: #Y aceptamos si todos son menores que la tolerancia
            Autovalores = [A[i][i] for i in range(n)] #Almacenamos los autovalores aqui
            return A,Autovalores,iteracion
    print(A)
    print(Q)
    print(Q.transpose() @ Q)
    return "Iteraciones máximas alcanzadas" #Error si no converge suficientemente rápido

In [100]:
A = np.array([[2, 1], 
             [3, 4]])

B = np.array([[3, 2],
             [3, 4]])  

C = np.array([[2, 3],
             [1, 4]])

D = np.array([[1, 1, 2],
             [2, 1, 1], 
             [1, 1, 3]])

E = np.array([[1, 1, 2],
             [2, 1, 3],
             [1, 1, 1]])

F = np.array([[2, 1, 2],
             [1, 1, 3],
             [1, 1, 1]])

G = np.array([[1, 1, 1, 2], 
             [2, 1, 1, 1],
             [3, 2, 1, 2],
             [2, 1, 1, 4]])

H = np.array([[1, 2, 1, 2],
             [2, 1, 1, 1], 
             [3, 2, 1, 2],
             [2, 1, 1, 4]])

Hagamos primero un ejemplo guiado sobre la matriz $A = \begin{bmatrix} 2 & 1 \\ 3 & 4\end{bmatrix}$, haciendo los primeros 3 pasos:

Primero imprimimos $A_0 = A$, después $Q_0$ y $R_0$, y después $A_1 = R_0 Q_0$, y seguimos. Obtenemos lo siguiente:

In [101]:
print(A)
A1 = A
for i in range(3):
    Q,R = QR_factor(A1)
    print(Q)
    print(R)
    A1=R@Q
    print(A1)

[[2 1]
 [3 4]]
[[ 0.5547002  -0.83205029]
 [ 0.83205029  0.5547002 ]]
[[3.60555128 3.88290137]
 [0.         1.38675049]]
[[ 5.23076923 -0.84615385]
 [ 1.15384615  0.76923077]]
[[ 0.97652377 -0.21540966]
 [ 0.21540966  0.97652377]]
[[ 5.35652011 -0.66058961]
 [ 0.          0.93344184]]
[[ 5.08847185 -1.79892761]
 [ 0.20107239  0.91152815]]
[[ 0.99922018 -0.03948446]
 [ 0.03948446  0.99922018]]
[[ 5.09244302 -1.76153358]
 [ 0.          0.98184702]]
[[ 5.01891864 -1.9612323 ]
 [ 0.0387677   0.98108136]]


### Paso 1:
$A_0 = A = \begin{bmatrix}
2 & 1 \\
3 & 4
\end{bmatrix} = Q_1\, R_1 = \begin{bmatrix}
0.5547002 & -0.83205029 \\
0.83205029 & 0.5547002
\end{bmatrix} \begin{bmatrix}
3.60555128 & 3.88290137 \\
0 & 1.38675049
\end{bmatrix}$
### Paso 2:
$A_1 = R_0 \, Q_0 = \begin{bmatrix}
5.23076923 & -0.84615385 \\
1.15384615 & 0.76923077
\end{bmatrix} = Q_2 \, R_2 = \begin{bmatrix}
0.97652377 & -0.21540966 \\
0.21540966 & 0.97652377
\end{bmatrix}\begin{bmatrix}
5.35652011 & -0.66058961 \\
0 & 0.93344184
\end{bmatrix}$
### Paso 3:
$A_2 = R_1 \, Q_1 = \begin{bmatrix}
5.08847185 & -1.79892761 \\
0.20107239 & 0.91152815
\end{bmatrix} = Q_2 \, R_2 =\begin{bmatrix}
0.99922018 & -0.03948446 \\
0.03948446 & 0.99922018
\end{bmatrix} \begin{bmatrix}
5.09244302 & -1.76153358 \\
0 & 0.98184702
\end{bmatrix}$
### Paso final:
Después de todo esto, obtenemos la matriz $A_3 = \begin{bmatrix}
5.01891864 & -1.9612323 \\
0.0387677 & 0.98108136
\end{bmatrix} $
En cada uno de los pasos, podemos apreciar como disminuye la componente inferior izquierda de la matriz, el el paso 1 disminuye un 38.4%, después un 17.4%, y por último un 19.3% aproximadamente. Si corremos el código para hallar los valores propios, obtenemos lo siguiente:

In [102]:
AA,Diag,i = QR_algor(A,tol=1e-4)
with np.printoptions(precision=6, suppress=True, formatter={'float': '{:0.5f}'.format}, linewidth=100):
    print("Matriz (casi) diagonal similar a A:")
    print(AA)
print(f"Autovalores de A:{Diag}, en {i} pasos")

Matriz (casi) diagonal similar a A:
[[5.00003 -1.99994]
 [0.00006 0.99997]]
Autovalores de A:[5.000030719292191, 0.9999692807078088], en 7 pasos


In [103]:
test = np.array([[1,0,0,0],[3,2,0,0],[1,1,6,0],[1,2,1,3]])
QR_algor(test)

(array([[ 6.00000000e+00,  9.99999857e-01,  1.58113892e+00,
          1.26491106e+00],
        [ 2.48486930e-12,  3.00000014e+00, -1.58113874e+00,
         -6.32455803e-01],
        [ 1.25511323e-19,  9.03867370e-08,  1.99999986e+00,
          2.99999994e+00],
        [ 4.08151698e-34,  5.02619715e-21, -8.65008504e-14,
          1.00000000e+00]]),
 [5.999999999999165, 3.000000142914779, 1.9999998570863093, 0.99999999999974],
 43)